presto-main/src/test/java/com/facebook/presto/operator/TestExchangeClient.java

```diff
@@ -329,6 +335,69 @@ public void testClose()
         assertStatus(clientStatus, location, "closed", "not scheduled");
     }
 
+    @Test
+    public void testInitialRequestLimit()
+    {
+        DataSize maxResponseSize = new DataSize(1, BYTE);
+        MockExchangeRequestProcessor processor = new MockExchangeRequestProcessor(maxResponseSize);
+        List<URI> locations = new ArrayList<>();
+        int initialRequestCount = 8;
+        int numLocations = initialRequestCount * 2;
+
+        // add pages
+        for (int ii = 0; ii < numLocations; ii++)
+        {
+            URI location = URI.create("http://localhost:" + (8080 + ii));
+            locations.add(location);
+
+            processor.addPage(location, createPage(1));
+            processor.addPage(location, createPage(2));
+            processor.addPage(location, createPage(3));
+
+            processor.setComplete(location);
+        }
+
+        try (ExchangeClient exchangeClient = new ExchangeClient(
+                new DataSize(8, BYTE),
+                maxResponseSize,
+                initialRequestCount,
+                1,
+                new Duration(1, MINUTES),
+                true,
+                new TestingHttpClient(processor, testingHttpClientExecutor),
+                scheduler,
+                new SimpleLocalMemoryContext(newSimpleAggregatedMemoryContext(), "test"),
+                pageBufferClientCallbackExecutor)) {
+
+            for (int ii = 0; ii < numLocations; ii++) {
+                exchangeClient.addLocation(locations.get(ii), TaskId.valueOf("taskid" + ii));
+            }
+            exchangeClient.noMoreLocations();
+            assertFalse(exchangeClient.isClosed());
+
+            long start = System.nanoTime();
+
+            // start fetching pages
+            exchangeClient.scheduleRequestIfNecessary();
```

Comment:
`addLocation` calls this internally. No need to call it explicitly.

Reply:
Removed